components/brave_extension/extension/brave_extension/content_element_picker.ts

```diff
@@ -0,0 +1,424 @@
+  }
+
+  const querySelectorNoExcept = (node: Element | null, selector: string): Element[] => {
+    if (node !== null) {
+      try {
+        let r = node.querySelectorAll(selector)
+        return Array.from(r)
+      } catch (e) { /* Deliberately left empty */ }
+    }
+    return []
+  }
+
+  if (builder.size() === 0 || querySelectorNoExcept(elem.parentElement, builder.toString()).length > 1) {
+    builder.addTag(tag)
+    if (querySelectorNoExcept(elem.parentElement, builder.toString()).length > 1) {
+      let index = 1
+      let sibling: Element | null = elem.previousElementSibling
+      while (sibling !== null) {
+        if (sibling.localName === tag) { index++ }
+        sibling = sibling.previousElementSibling
+      }
+      builder.addRule({
+        type: Selector.NthOfType,
+        value: index
+      })
+    }
+  }
+  return builder
+}
+
+const elementPickerOnKeydown = (event: KeyboardEvent): void => {
+  if (event.key === 'Escape') {
+    event.stopPropagation()
+    event.preventDefault()
+    // tslint:disable-next-line:no-use-before-declare
+    quitElementPicker()
+  }
+}
+
+const elementPickerViewportChanged = () => {
+  // tslint:disable-next-line:no-use-before-declare
+  recalculateAndSendTargets(targetedElems)
+}
+
+const quitElementPicker = () => {
+  if (pickerFrame !== null) {
+    document.documentElement.removeChild(pickerFrame)
+  }
+
+  // Tear down element picker listeners
+  document.removeEventListener('keydown', elementPickerOnKeydown, true)
+  document.removeEventListener('resize', elementPickerViewportChanged)
+  document.removeEventListener('scroll', elementPickerViewportChanged)
+}
+
+const launchElementPicker = () => {
+  pickerFrame = document.createElement('iframe')
+  // "src" is a web accessible resource since the URI is chrome-extension://.
+  // This ensures a malicious page cannot modify the iframe contents.
+  pickerFrame.src = chrome.runtime.getURL('elementPicker.html')
+  const pickerCSSStyle: string = [
+    'background: transparent',
+    'border: 0',
+    'border-radius: 0',
+    'box-shadow: none',
+    'display: block',
+    'height: 100%',
+    'left: 0',
+    'margin: 0',
+    'max-height: none',
+    'max-width: none',
+    'opacity: 1',
+    'outline: 0',
+    'padding: 0',
+    'pointer-events: auto',
+    'position: fixed',
+    'top: 0',
+    'visibility: visible',
+    'width: 100%',
+    'z-index: 2147483647',
+    ''
+  ].join(' !important;')
+  // TODO(keur): To harden this we can use chrome.tabs.insertCSS(). Not a
+  // big priority since the page can always remove the iframe if it wants.
+  pickerFrame.style.cssText = pickerCSSStyle
+
+  // We don't append to the body because we are setting the frame's
+  // width and height to be 100%. Prevents the picker from only being
+  // able to hover the iframe.
+  document.documentElement.appendChild(pickerFrame)
+
+  // Setup listeners to assist element picker
+  document.addEventListener('keydown', elementPickerOnKeydown, true)
```

Comment:
why true here?

Reply:
i can change this. i don't think it matters that much. i just wanted to ensure that the `Esc` event to teardown the picker isn't eaten up by some other listener with `e.preventDefault()`.